akka-projection-core/src/test/scala/akka/projection/internal/metrics/InternalProjectionStateMetricsSpec.scala

```diff
@@ -0,0 +1,377 @@
+
+  /**
+   * @param erroredOffsets a stack of errors. Must be ordered. Each item is an offset which, when observed will
+   *                       trigger an error and then be removed from the stack. To fail an item multiple times
+   *                       add its offset repeatedly. Uses `Int` instead of `Long` for convenience.
+   */
+  def singleWithErrors(erroredOffsets: Int*): () => Handler[Envelope] = {
+    var nextProcessStrategy = ProcessStrategy(erroredOffsets.map { _.toLong }.toList)
+    () =>
+      new Handler[Envelope] {
+        override def process(envelope: Envelope): Future[Done] = {
+          nextProcessStrategy match {
+            case SomeFailures(nextFail :: tail) if (nextFail == envelope.offset) =>
+              nextProcessStrategy = SomeFailures(tail)
+              throw TelemetryException
+            case _ => Future.successful(Done)
+          }
+        }
+      }
+  }
+
+  val grouped = groupedWithErrors()
+
+  /**
+   * @param erroredOffsets a stack of errors. Must be ordered. Each item is an offset which, when observed will
+   *                       trigger an error and then be removed from the stack. To fail an item multiple times
+   *                       add its offset repeatedly. Uses `Int` instead of `Long` for convenience.
+   */
+  def groupedWithErrors(erroredOffsets: Int*): () => Handler[immutable.Seq[Envelope]] = {
+    var nextProcessStrategy = ProcessStrategy(erroredOffsets.map { _.toLong }.toList)
+    () =>
+      new Handler[immutable.Seq[Envelope]] {
+        override def process(envelopes: immutable.Seq[Envelope]): Future[Done] = {
+          nextProcessStrategy match {
+            case SomeFailures(nextFail :: tail) if (envelopes.map { _.offset }.contains(nextFail)) =>
+              nextProcessStrategy = SomeFailures(tail)
+              Future.failed(TelemetryException)
+            case _ =>
+              Future.successful(Done)
+          }
+        }
+      }
+  }
+
+  val flow = flowWithErrors()
+
+  /**
+   * @param erroredOffsets a stack of errors. Must be ordered. Each item is an offset which, when observed will
+   *                       trigger an error and then be removed from the stack. To fail an item multiple times
+   *                       add its offset repeatedly. Uses `Int` instead of `Long` for convenience.
+   */
+  def flowWithErrors(erroredOffsets: Int*) = {
+    var nextProcessStrategy = ProcessStrategy(erroredOffsets.map { _.toLong }.toList)
+    FlowWithContext[Envelope, ProjectionContext]
+      .map { envelope =>
+        nextProcessStrategy match {
+          case SomeFailures(nextFail :: tail) if (envelope.offset == nextFail) =>
+            nextProcessStrategy = SomeFailures(tail)
+            throw TelemetryException
+          case _ =>
+            Done
+        }
+      }
+  }
+
+}
+
+object InMemInstruments {
+  private val instrumentMap = new ConcurrentHashMap[ProjectionId, InMemInstruments]()
```

Comment:
One global storage with instruments. The instruments in these tests survive any death of `InMemTelemetry` instances so we can run assertions at the end of a test.

Reply:
Each test must use a unique `ProjectionId` so they use dedicated instruments.

Reply:
Nice! Looks cleaner too.